File: backend/src/routes/order.ts
import { CreateOrderPost, OrderUpdate } from "@/interface/api/Order";
import { responseHandler } from "@/services/Handler";
import {
	CreateOrder,
	DeleteOrder,
	GetHistoryOrders,
	GetOrder,
	GetOrders,
	OrderDone,
	UpdateOrder,
} from "@/services/Order";
import express from "express";
import { ObjectId } from "mongoose";

const orderRoute = express.Router();

orderRoute.post("/", async (req, res) => {
	const data: CreateOrderPost = req.body;
	return responseHandler(res, await CreateOrder(req, data));
});

orderRoute.delete("/delete", async (req, res) => {
	const order_id: ObjectId = req.body.order_id;
	return responseHandler(res, await DeleteOrder(req, order_id));
});

orderRoute.get("/", async (req, res) => {
	return responseHandler(res, await GetOrders(req));
});

orderRoute.get("/history", async (req, res) => {
	return responseHandler(res, await GetHistoryOrders(req));
});

orderRoute.get("/:order_id", async (req, res) => {
	return responseHandler(res, await GetOrder(req));
});

orderRoute.patch("/update", async (req, res) => {
	const data: OrderUpdate = req.body;
	return responseHandler(res, await UpdateOrder(req, data));
});

orderRoute.patch("/done", async (req, res) => {
	const order_id: ObjectId = req.body.order_id;
	return responseHandler(res, await OrderDone(req, order_id));
});

export default orderRoute;

File: sheepper/lib/services/api/order.dart
import 'package:sheepper/models/order.dart';
import 'package:sheepper/models/response/error_response.dart';
import 'package:sheepper/models/response/info_response.dart';
import 'package:sheepper/services/dio.dart';
import 'package:sheepper/services/share_preference.dart';

class OrderApi {
  static Future<dynamic> getBigOrder() async {
    DioInstance.dio.options.headers["authorization"] =
        "Bearer " + SharePreference.prefs.getString("token").toString();
    final response = await DioInstance.dio.get("/order");
    if (response.statusCode != 200 && response.statusCode != 201) {
      return ErrorResponse.fromJson(response.data);
    } else {
      return InfoResponse.fromJson(response.data);
    }
  }

  static Future<dynamic> getOrder(String orderId) async {
    DioInstance.dio.options.headers["authorization"] =
        "Bearer " + SharePreference.prefs.getString("token").toString();
    final response = await DioInstance.dio.get("/order/$orderId");
    if (response.statusCode != 200 && response.statusCode != 201) {
      return ErrorResponse.fromJson(response.data);
    } else {
      return InfoResponse.fromJson(response.data);
    }
  }

  static Future<dynamic> getOldOrder(String orderId) async {
    DioInstance.dio.options.headers["authorization"] =
        "Bearer " + SharePreference.prefs.getString("token").toString();
    final response = await DioInstance.dio.get("/order/history/$orderId");
    if (response.statusCode != 200 && response.statusCode != 201) {
      return ErrorResponse.fromJson(response.data);
    } else {
      return InfoResponse.fromJson(response.data);
    }
  }

  static Future<dynamic> updateOrder(List<OrderDetailModel> order) async {
    DioInstance.dio.options.headers["authorization"] =
        "Bearer " + SharePreference.prefs.getString("token").toString();

    final response = await DioInstance.dio.patch("/order/update",
        data: {"order_id": order[0].order_id, "detail": order});
    if (response.statusCode != 200 && response.statusCode != 201) {
      return ErrorResponse.fromJson(response.data);
    } else {
      return InfoResponse.fromJson(response.data);
    }
  }

  static Future<dynamic> deleteOrder(String orderId) async {
    DioInstance.dio.options.headers["authorization"] =
        "Bearer " + SharePreference.prefs.getString("token").toString();
    final response = await DioInstance.dio.delete("/order/$orderId");
    if (response.statusCode != 200 && response.statusCode != 201) {
      return ErrorResponse.fromJson(response.data);
    } else {
      return InfoResponse.fromJson(response.data);
    }
  }

  static Future<dynamic> doneOrder(String orderId) async {
    DioInstance.dio.options.headers["authorization"] =
        "Bearer " + SharePreference.prefs.getString("token").toString();
    final response = await DioInstance.dio.patch("/order/done/$orderId");
    if (response.statusCode != 200 && response.statusCode != 201) {
      return ErrorResponse.fromJson(response.data);
    } else {
      return InfoResponse.fromJson(response.data);
    }
  }


  static Future<dynamic> getBigHistoryOrder() async {
    DioInstance.dio.options.headers["authorization"] =
        "Bearer " + SharePreference.prefs.getString("token").toString();
    final response = await DioInstance.dio.get("/order/history");
    if (response.statusCode != 200 && response.statusCode != 201) {
      return ErrorResponse.fromJson(response.data);
    } else {
      return InfoResponse.fromJson(response.data);
    }
  }
}
